cirq-core/cirq/transformers/dynamical_decoupling.py

```diff
@@ -98,114 +110,75 @@ def _is_single_qubit_gate_moment(moment: 'cirq.Moment') -> bool:
 
 
+def _merge_single_qubit_ops_to_phxz(
+    q: 'cirq.Qid', ops: Tuple['cirq.Operation', ...]
+) -> 'cirq.Operation':
+    """Merges [op1, op2, ...] and returns an equivalent op"""
+    if len(ops) == 1:
+        return ops[0]
+    matrices = [cirq.unitary(op) for op in reversed(ops)]
+    product = reduce(np.matmul, matrices)
+    gate: Optional['cirq.Gate'] = single_qubit_decompositions.single_qubit_matrix_to_phxz(product)
+    if gate is None:
+        # gate is close to Identity.
+        gate = cirq.I
+    return gate.on(q)
+
+
+def _merge_single_qubit_ops_of_two_moments(
+    m1: 'cirq.Moment', m2: 'cirq.Moment'
+) -> Tuple['cirq.Moment', ...]:
+    for q in m1.qubits | m2.qubits:
+        op1 = m1.operation_at(q)
+        op2 = m2.operation_at(q)
+        if op1 is not None and op2 is not None:
+            if any(
+                not (_is_single_qubit_operation(op) and cirq.has_unitary(op)) for op in [op1, op2]
+            ):
+                return (m1, m2)
+    ops: set['cirq.Operation'] = set()
+    for q in m1.qubits | m2.qubits:
+        op12 = [op for op in [m.operation_at(q) for m in [m1, m2]] if op is not None]
+        ops.add(_merge_single_qubit_ops_to_phxz(q, tuple(op12)))
+    return tuple([cirq.Moment(ops)])
+
+
 def _calc_pulled_through(
     moment: 'cirq.Moment', input_pauli_ops: 'cirq.PauliString'
 ) -> 'cirq.PauliString':
-    """Calculates the pulled_through after pulling through moment with the input.
-
-    We assume that the moment is Clifford here. Then, pulling through is essentially
-      decomposing a matrix into Pauli operations on each qubit.
-    """
+    """Calculates the pulled_through after pulling through clifford ops in moment."""
     pulled_through: 'cirq.PauliString' = cirq.PauliString()
-    for affected_q, combined_op_in_pauli in input_pauli_ops.items():
+    for affected_q, combined_gate_in_pauli in input_pauli_ops.items():
         op_at_moment = moment.operation_at(affected_q)
-        if op_at_moment is None:
-            pulled_through *= combined_op_in_pauli.on(affected_q)
+        if op_at_moment is None:  # keep the pauli op through empty op
+            pulled_through *= combined_gate_in_pauli.on(affected_q)
             continue
-        prev_circuit = cirq.Circuit(cirq.Moment(op_at_moment))
-        new_circuit = cirq.Circuit(
-            cirq.Moment(combined_op_in_pauli.on(affected_q)), cirq.Moment(op_at_moment)
-        )
-        qubit_order = op_at_moment.qubits
-        pulled_through_pauli_ops = unitary_to_pauli_string(
-            prev_circuit.unitary(qubit_order=qubit_order)
-            @ new_circuit.unitary(qubit_order=qubit_order).conj().T
-        )
-        if pulled_through_pauli_ops is not None:
-            for qid, gate in enumerate(pulled_through_pauli_ops):
-                pulled_through *= gate.on(qubit_order[qid])
+        if _is_clifford_op(op_at_moment):
+            prev_circuit = cirq.Circuit(cirq.Moment(op_at_moment))
```

Comment:
this is inefficient. looks like you only need the unitary which you can get using `cirq.protocols.unitary_protocol.unitary(op)`